server-spi-private/src/main/java/org/keycloak/userprofile/DefaultUserProfile.java

```diff
@@ -0,0 +1,134 @@
+ *  * and other contributors as indicated by the @author tags.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License");
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  * http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.keycloak.userprofile;
+
+import java.util.AbstractMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import org.keycloak.models.UserModel;
+
+/**
+ * <p>The default implementation for {@link UserProfile}. Should be reused as much as possible by the different implementations
+ * of {@link UserProfileProvider}.
+ *
+ * <p>This implementation is not specific to any user profile implementation.
+ *
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public class DefaultUserProfile implements UserProfile {
+
+    private final Function<DefaultAttributes, UserModel> userSupplier;
+    private final DefaultAttributes attributes;
+    private boolean validated;
+    private UserModel user;
+
+    public DefaultUserProfile(DefaultAttributes attributes, Function<DefaultAttributes, UserModel> userSupplier, UserModel user) {
+        this.userSupplier = userSupplier;
+        this.attributes = attributes;
+        this.user = user;
+    }
+
+    @Override
+    public void validate() {
+        ValidationException validationException = new ValidationException();
+
+        for (String attributeName : attributes.keySet()) {
+            this.attributes.validate(attributeName,
+                    (attribute, message) -> validationException.addError(new ValidationException.Error(attribute, message)));
+        }
+
+        if (validationException.hasError()) {
+            throw validationException;
+        }
+
+        validated = true;
+    }
+
+    @Override
+    public UserModel create() throws ValidationException {
+        if (user != null) {
+            throw new RuntimeException("User already created");
+        }
+
+        user = userSupplier.apply(this.attributes);
+
+        return updateInternal(false);
+    }
+
+    @Override
+    public void update(boolean removeAttributes, BiConsumer<String, UserModel>... changeListener) {
+        updateInternal(removeAttributes, changeListener);
+    }
+
+    private UserModel updateInternal(boolean removeAttributes, BiConsumer<String, UserModel>... changeListener) {
```

Comment:
Make this protected please to allow customization

Reply:
This is really internal code. Customizations to update should be done through a specific hook and better API, which is not yet here.

Do you have any examples from your use cases on how you want to customize user updates so that we can think about something better?

Reply:
I expect UserProfile is SPI, so I expect it will be used for customizations/extensions, and extending current code is easiest way for customizations. private methods make extensions much harder and push you to copy whole code to your extension, which is not good for long term maintenance and upgrades to newer keycloak versions (which we need).
Our use case is that we have some attributes which need to be configured into user profiles in JSON config, shown in registration and user update forms, validated to decide that user matches user profile level, but they are not persisted in UserModel but in another external system (REST API). Communication with the external system should be performed when really necessary (when current user profile requires given attributes). So for me it seems that simplest way is to customize UserProfileProvider to handle these attributes, but I do not want to rewrite too much code

Reply:
Sorry @velias, but it is just a matter of changing private to protected and not copy/paste. Nobody is proposing copy/paste, but to think about the best way to allow customizations with real use cases and a proper solution.

Overriding that method is definitely not the best way of extending and solving your use case. If you want less code we could provide specific hooks or event listeners for deciding whether an attribute should be persisted based on some logic.
By overriding this method you are going against your goal of not having to write too much code as you are going to need to still have some similar code as in this method.

Again, we should have proper hooks. We can either consider this requirement now or perhaps work on it later once we have the core changes in. It should just be a matter of making this part easier to customize.

Reply:
Right, I agree, it is not easy to do it correctly now. I'd say once we finish basic functionality of the SPI I'll start to re-implement our code based on it an we will see what is necessary and we can extend provider internals then.